File: app/src/main/java/com/mirogal/cocktail/presentation/ui/searchlist/CocktailSearchListActivity.java
package com.mirogal.cocktail.presentation.ui.searchlist;

import android.content.Intent;
import android.content.res.Configuration;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.LinearLayout;

import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.SearchView;
import androidx.appcompat.widget.Toolbar;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.mirogal.cocktail.R;
import com.mirogal.cocktail.data.database.entity.CocktailDbEntity;
import com.mirogal.cocktail.data.repository.NetworkState;
import com.mirogal.cocktail.presentation.ui.constant.IntentTag;
import com.mirogal.cocktail.presentation.ui.detail.CocktailDetailActivity;
import com.mirogal.cocktail.presentation.ui.util.GridSpaceItemDecoration;
import com.mirogal.cocktail.presentation.viewmodel.searchlist.CocktailSearchListViewModel;

public class CocktailSearchListActivity extends AppCompatActivity
        implements ListAdapter.OnItemClickListener {

    // TODO Dagger integration

    private CocktailSearchListViewModel viewModel;

    private RecyclerView recyclerView;
    private GridLayoutManager layoutManager;
    private GridSpaceItemDecoration itemDecoration;
    private ListAdapter listAdapter;
    private LinearLayout layoutEmpty;
    private LinearLayout layoutStart;

    private SearchView searchView;
    private MenuItem searchMenuItem;
    private String requestQuery;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_cocktail_search_list);

        viewModel = new ViewModelProvider(this).get(CocktailSearchListViewModel.class);

        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        int listColumn;
        int orientation = this.getResources().getConfiguration().orientation;
        if (orientation == Configuration.ORIENTATION_PORTRAIT) {
            listColumn = 2;
        } else {
            listColumn = 3;
        }

        recyclerView = findViewById(R.id.recyclerView);
        layoutEmpty = findViewById(R.id.layout_search_list_empty);
        layoutStart = findViewById(R.id.layout_search_list_start);

        layoutManager = new GridLayoutManager(this, listColumn);
        recyclerView.setLayoutManager(layoutManager);

        int spaceInPixel = getResources().getDimensionPixelSize(R.dimen.padding_horizontal);
        itemDecoration = new GridSpaceItemDecoration(listColumn, spaceInPixel, true, 0);
        recyclerView.addItemDecoration(itemDecoration);

        listAdapter = new ListAdapter(this, this, R.layout.item_cocktail);
//        viewModel.cocktailList.observe(this, pagedList -> {
//            try {
//                listAdapter.submitList(pagedList);
//            } catch (Exception ignored) {}
//        });
//        recyclerView.setAdapter(listAdapter);

        viewModel.getNetworkStatus().observe(this, status -> {
            if (status == NetworkState.EMPTY) {
                showEmpty();
            } else {
                showData();
            }
        });

        viewModel.getRequestQuery().observe(this, query -> {
            CocktailSearchListActivity.this.requestQuery = query;
            if (query == null || query.isEmpty()) {
                showStart();
            }
        });
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.cocktail_search_list_function, menu);
        searchMenuItem = menu.findItem(R.id.action_search);
        searchView = (SearchView) searchMenuItem.getActionView();
        searchView.setIconifiedByDefault(false); // set inner icon
        searchView.setFocusable(true);
        searchView.setIconified(false);
        searchView.requestFocusFromTouch();
        if (requestQuery != null) {
            searchView.setQuery(requestQuery, false);
        }
        searchView.setOnQueryTextListener(
                new SearchView.OnQueryTextListener() {
                    @Override
                    public boolean onQueryTextSubmit(String s) {
                        return false;
                    }

                    @Override
                    public boolean onQueryTextChange(String s) {
                        viewModel.setRequestQuery(s);
                        return false;
                    }
                }
        );
        return true;
    }

    @Override
    public void onItemClick(CocktailDbEntity cocktail) {
        viewModel.saveCocktail(cocktail);
        openCocktailDetailActivity(cocktail);
    }

    private void openCocktailDetailActivity(CocktailDbEntity cocktail) {
        Intent intent = new Intent(CocktailSearchListActivity.this, CocktailDetailActivity.class);
        intent.putExtra(IntentTag.COCKTAIL_ENTITY.toString(), cocktail);
        startActivity(intent);
    }

    private void showData() {
        if (recyclerView.getVisibility() == View.INVISIBLE) {
            recyclerView.setVisibility(View.VISIBLE);
            layoutEmpty.setVisibility(View.INVISIBLE);
            layoutStart.setVisibility(View.INVISIBLE);
        }
    }

    private void showEmpty() {
        if (layoutEmpty.getVisibility() == View.INVISIBLE) {
            recyclerView.setVisibility(View.INVISIBLE);
            layoutEmpty.setVisibility(View.VISIBLE);
            layoutStart.setVisibility(View.INVISIBLE);
        }
    }

    private void showStart() {
        if (layoutStart.getVisibility() == View.INVISIBLE) {
            recyclerView.setVisibility(View.INVISIBLE);
            layoutEmpty.setVisibility(View.INVISIBLE);
            layoutStart.setVisibility(View.VISIBLE);
        }
    }
}

File: app/src/main/java/com/mirogal/cocktail/ui/searchlist/CocktailSearchListActivity.kt
package com.mirogal.cocktail.ui.searchlist

import android.content.Intent
import android.content.res.Configuration
import android.os.Bundle
import android.view.Menu
import android.view.View
import androidx.appcompat.widget.SearchView
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.paging.PagedList
import androidx.recyclerview.widget.GridLayoutManager
import com.mirogal.cocktail.R
import com.mirogal.cocktail.ui.base.BaseActivity
import com.mirogal.cocktail.data.database.entity.CocktailDbEntity
import com.mirogal.cocktail.data.repository.NetworkState
import com.mirogal.cocktail.ui.constant.IntentTag
import com.mirogal.cocktail.ui.detail.CocktailDetailActivity
import com.mirogal.cocktail.ui.util.GridSpaceItemDecoration
import kotlinx.android.synthetic.main.activity_cocktail_save_list.*
import kotlinx.android.synthetic.main.content_cocktail_save_list.*
import kotlinx.android.synthetic.main.layout_save_list_empty.*
import kotlinx.android.synthetic.main.layout_search_list_start.*

class CocktailSearchListActivity : BaseActivity(), ListAdapter.OnItemClickListener {

    private lateinit var viewModel: ViewModel
    private var requestQuery: String? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_cocktail_search_list)

        setSupportActionBar(toolbar)

        val listColumn = when (this.resources.configuration.orientation) {
            Configuration.ORIENTATION_PORTRAIT -> 2
            Configuration.ORIENTATION_LANDSCAPE -> 3
            else -> 1
        }
        recyclerView.layoutManager = GridLayoutManager(this, listColumn)

        val spaceInPixel = resources.getDimensionPixelSize(R.dimen.padding_horizontal)
        val itemDecoration = GridSpaceItemDecoration(listColumn, spaceInPixel, true, 0)
        recyclerView.addItemDecoration(itemDecoration)

        val listAdapter = ListAdapter(this, this, R.layout.item_cocktail)
        viewModel = ViewModelProvider(this).get(ViewModel::class.java)
        viewModel.cocktailList.observe(this, Observer { pagedList: PagedList<CocktailDbEntity?> ->
            try {
                listAdapter.submitList(pagedList)
            } catch (ignored: Exception) {
            }
        })
        recyclerView.adapter = listAdapter

        viewModel.networkStatus.observe(this, Observer { status: NetworkState.Status ->
            if (status == NetworkState.EMPTY) {
                showEmpty()
            } else {
                showData()
            }
        })

        viewModel.getRequestQuery().observe(this, Observer { query: String? ->
            requestQuery = query
            if (!(query != null && query.isNotEmpty())) {
                showStart()
            }
        })
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.cocktail_search_list_function, menu)
        val searchMenuItem = menu.findItem(R.id.action_search)
        val searchView = searchMenuItem.actionView as SearchView
        searchView.setIconifiedByDefault(false) // set inner icon
        searchView.isFocusable = true
        searchView.isIconified = false
        searchView.requestFocusFromTouch()
        if (requestQuery != null) {
            searchView.setQuery(requestQuery, false)
        }
        searchView.setOnQueryTextListener(
                object : SearchView.OnQueryTextListener {
                    override fun onQueryTextSubmit(s: String): Boolean {
                        return false
                    }

                    override fun onQueryTextChange(s: String): Boolean {
                        viewModel.setRequestQuery(s)
                        return false
                    }
                }
        )
        return true
    }

    override fun onItemClick(cocktail: CocktailDbEntity?) {
        viewModel.saveCocktail(cocktail)
        openCocktailDetailActivity(cocktail)
    }

    private fun openCocktailDetailActivity(cocktail: CocktailDbEntity?) {
        val intent = Intent(this@CocktailSearchListActivity, CocktailDetailActivity::class.java)
        intent.putExtra(IntentTag.COCKTAIL_ENTITY.toString(), cocktail)
        startActivity(intent)
    }

    private fun showData() {
        if (recyclerView.visibility == View.INVISIBLE) {
            recyclerView.visibility = View.VISIBLE
            layoutEmpty.visibility = View.INVISIBLE
            layoutStart.visibility = View.INVISIBLE
        }
    }

    private fun showEmpty() {
        if (layoutEmpty.visibility == View.INVISIBLE) {
            recyclerView.visibility = View.INVISIBLE
            layoutEmpty.visibility = View.VISIBLE
            layoutStart.visibility = View.INVISIBLE
        }
    }

    private fun showStart() {
        if (layoutStart.visibility == View.INVISIBLE) {
            recyclerView.visibility = View.INVISIBLE
            layoutEmpty.visibility = View.INVISIBLE
            layoutStart.visibility = View.VISIBLE
        }
    }

}